cirq/ops/linear_operator.py

```diff
@@ -0,0 +1,314 @@
+        """
+        matrix = self._matrix_()
+        if matrix is None:
+            matrix = operator_spaces.reconstruct_from_expansion(
+                self._pauli_expansion_(), operator_spaces.PAULI_BASIS)
+        exponent = matrix * np.log(other)
+        matrix = scipy.linalg.expm(exponent)
+
+        pauli_expansion = self._pauli_expansion_()
+        if pauli_expansion is not None:
+            pauli_expansion = operator_spaces.expand_in_basis(
+                matrix, operator_spaces.PAULI_BASIS)
+
+        return LinearOperator(matrix, pauli_expansion)
+
+    def exp(self) -> 'AbstractLinearOperator':
+        """Computes operator power of number e."""
+        return self.__rpow__(np.e)
+
+    def polar_decomposition(self) -> Tuple['AbstractLinearOperator',
+                                           'AbstractLinearOperator']:
+        """Computes (right) polar decomposition of self."""
+        matrix = self._matrix_()
+        if matrix is None:
+            pauli_expansion = self._pauli_expansion_()
+            matrix = operator_spaces.reconstruct_from_expansion(
+                pauli_expansion, operator_spaces.PAULI_BASIS)
+        u_matrix, p_matrix = scipy.linalg.polar(matrix)
+
+        u_pauli_expansion, p_pauli_expansion = None, None
+        if self._pauli_expansion_() is not None:
+            u_pauli_expansion = operator_spaces.expand_in_basis(
+                u_matrix, operator_spaces.PAULI_BASIS)
+            p_pauli_expansion = operator_spaces.expand_in_basis(
+                p_matrix, operator_spaces.PAULI_BASIS)
+
+        return (LinearOperator(u_matrix, u_pauli_expansion),
+                LinearOperator(p_matrix, p_pauli_expansion))
+
+    def unitary_factor(self) -> 'AbstractLinearOperator':
+        """Computes (right) unitary factor of self."""
+        u, _ = self.polar_decomposition()
+        return u
+
+
+class UnitaryMixin(AbstractLinearOperator):
+    """Mixin which equips a quantum gate with linear algebraic operations.
+
+    To enable linear algebraic operations involving a gate, the gate must
+    override one or both of _pauli_expansion_(), _unitary_().
+    """
+    def _n_dim_(self) -> Optional[int]:
+        if isinstance(self, gate_features.SingleQubitGate):
+            return 2
+        if isinstance(self, gate_features.TwoQubitGate):
+            return 4
+        if isinstance(self, gate_features.ThreeQubitGate):
+            return 8
+        return None
+
+    def _matrix_(self) -> Optional[np.ndarray]:
+        """Returns the matrix of the unitary operator."""
+        return protocols.unitary(self, None)
+
+    def _pauli_expansion_(self) -> Optional[np.ndarray]:
+        """Returns the expansion of the unitary operator in the Pauli basis.
+
+        Gates which know the expansion, should override.
+        """
+        pass
+
+
+class LinearOperator(AbstractLinearOperator):
+    """General linear operator on a vector space.
```

Comment:
Re requirement for a matrix:
I do assume we're only considering finite dimensional vector spaces. Updated comment.

Re requirement for a Pauli expansion:
I do permit complex coefficients in the expansions. Note that while real linear combinations of Paulis generate the space of Hermitian operators, complex ones generate the entire space of all linear operators on a Hilbert space. You can see this easily by counting free parameters. For example `2 * a\dagger` is just X - 1j * Y which incidentally is used as test input in `test_operator_power()`.

That said, I think it's ok to introduce new internal representations that aren't general enough to describe all operators since any given instance of an operator is allowed a choice as to which representations to support.

Reply:
There are two different questions here:
1. What set of operators is being modeled (e.g. positive, hermitian, general...)?
2. What representation of the operators is being used (e.g. standard basis/matrix, Pauli expansion)?

In the comment, I meant "general" w.r.t. question 1, i.e. to signify lack of restriction to any particular subset of L(V) (space of linear operators on V). Not "general" as a synonym for "basis-independent".

Initially, I considered modeling these two notions separately. I had a group of classes to model operators in basis-independent way and a second group of classes to host all representation/basis-dependent logic. Eventually, I decided against this separation to avoid proliferation of classes. We can go back to that approach if we do in fact end up using many representations. I'd say that more than 3-4 bases would make the sheer size of the classes in the present monolithic approach ripe for break-up.

I'll clarify the comment tomorrow. Let me know if you have specific suggestions about what to include or how to rename the classes.